ruma-events-macros/src/event_enum.rs

```diff
@@ -572,85 +528,93 @@ mod kw {
     syn::custom_keyword!(events);
 }
 
-pub enum EnumKind {
+pub enum EventKindVariation {
+    Full,
+    Stub,
+    Stripped,
+    Redacted,
+    RedactedStub,
+    RedactedStripped,
+}
+
+pub enum EventKind {
     Basic(Ident),
     Ephemeral(Ident),
     Message(Ident),
     State(Ident),
     ToDevice(Ident),
 }
 
-impl EnumKind {
-    /// Return the span of this ident.
-    fn span(&self) -> Span {
-        match &self {
-            Self::Basic(i)
-            | Self::Ephemeral(i)
-            | Self::Message(i)
-            | Self::State(i)
-            | Self::ToDevice(i) => i.span(),
-        }
-    }
-    /// `[Kind]Event`
-    fn to_full_event_struct(&self) -> Ident {
-        match &self {
-            Self::Basic(i)
-            | Self::Ephemeral(i)
-            | Self::Message(i)
-            | Self::State(i)
-            | Self::ToDevice(i) => format_ident!("{}Event", i),
+impl EventKind {
+    fn is_state(&self) -> bool {
+        if let Self::State(_) = self {
+            true
+        } else {
+            false
         }
     }
 
-    /// `Any[Kind]Event`
-    fn to_full_event_enum(&self) -> Ident {
-        match self {
-            Self::Basic(i)
-            | Self::Ephemeral(i)
-            | Self::Message(i)
-            | Self::State(i)
-            | Self::ToDevice(i) => format_ident!("Any{}Event", i),
+    fn is_message(&self) -> bool {
+        if let Self::Message(_) = self {
+            true
+        } else {
+            false
         }
     }
 
-    /// `Any[Kind]EventStub`
-    fn to_event_stub_enum(&self) -> Option<Ident> {
-        match self {
-            Self::Ephemeral(i) | Self::Message(i) | Self::State(i) => {
-                Some(format_ident!("Any{}EventStub", i))
+    fn to_event_ident(&self, var: &EventKindVariation) -> Option<Ident> {
+        use EventKindVariation::*;
+
+        match var {
```

Comment:
You don't have to nest matches here, you can do

```rust
match (self, var) {
    // all `EventKind`s are valid event structs
    (_, Full) => Some(...),
    // this one would be so much nicer with feature(or_patterns) 🙄
    // https://github.com/rust-lang/rust/issues/54883
    (Self::Ephemeral(i), Stub) |
    (Self::State(i), Stub) |
    (Self::Message(i), Stub) => Some(...),
    // ...
    _ => None
}
```

Reply:
Yeah I went back and forth but really probably in 99% of situations flat is better. `or_patterns` will be SO nice for stuff like this. I'll flatten it out.